coderd/workspaces_test.go

```diff
@@ -426,47 +426,441 @@ func TestWorkspace(t *testing.T) {
 
-		org, err := client.Organization(ctx, ws.OrganizationID)
-		require.NoError(t, err)
-		require.Equal(t, ws.OrganizationName, org.Name)
+				ctx := testutil.Context(t, testutil.WaitLong)
+
+				// Check presets
+				presets, err := client.TemplateVersionPresets(ctx, version.ID)
+				require.NoError(t, err)
+				require.Equal(t, len(tc.presets), len(presets))
+
+				if len(tc.presets) > 0 {
+					// Verify preset names and parameters
+					for i, preset := range presets {
+						require.Equal(t, tc.presets[i].Name, preset.Name)
+
+						// Check if the preset should have parameters
+						if tc.presets[i].Parameters != nil {
+							// Verify that the preset has the expected parameters
+							for _, param := range tc.presets[i].Parameters {
+								found := false
+								for _, presetParam := range preset.Parameters {
+									if param.Name == presetParam.Name {
+										require.Equal(t, param.Value, presetParam.Value,
+											"Parameter %s should have value %s", param.Name, param.Value)
+										found = true
+										break
+									}
+								}
+								require.True(t, found, "Parameter %s should be present in preset", param.Name)
+							}
+						}
+					}
+				}
+
+				// Create workspace with or without preset
+				var workspace codersdk.Workspace
+				if tc.selectedPresetIndex >= 0 {
+					// Use the selected preset
+					workspace = coderdtest.CreateWorkspace(t, client, template.ID, func(request *codersdk.CreateWorkspaceRequest) {
+						request.TemplateVersionPresetID = presets[tc.selectedPresetIndex].ID
+					})
+				} else {
+					workspace = coderdtest.CreateWorkspace(t, client, template.ID)
+				}
+
+				// Verify workspace details
+				authz.Reset() // Reset all previous checks done in setup.
+				ws, err := client.Workspace(ctx, workspace.ID)
+				authz.AssertChecked(t, policy.ActionRead, ws)
+				require.NoError(t, err)
+				require.Equal(t, user.UserID, ws.LatestBuild.InitiatorID)
+				require.Equal(t, codersdk.BuildReasonInitiator, ws.LatestBuild.Reason)
+
+				// Check preset ID if expected
+				if tc.selectedPresetIndex >= 0 {
+					require.NotNil(t, ws.LatestBuild.TemplateVersionPresetID)
+					require.Equal(t, presets[tc.selectedPresetIndex].ID, *ws.LatestBuild.TemplateVersionPresetID)
+
+					// If the selected preset has parameters and there are template version parameters,
+					// verify that only matching parameters were applied
+					if tc.presets[tc.selectedPresetIndex].Parameters != nil && len(tc.templateVersionParameters) > 0 {
+						builds, err := client.WorkspaceBuilds(ctx, codersdk.WorkspaceBuildsRequest{
+							WorkspaceID: ws.ID,
+						})
+						require.NoError(t, err)
+						require.Equal(t, 1, len(builds))
+						parameters, err := client.WorkspaceBuildParameters(ctx, builds[0].ID)
+						require.NoError(t, err)
+						parametersSetByPreset := 0
+						for _, param := range parameters {
+							for _, presetParam := range tc.presets[tc.selectedPresetIndex].Parameters {
```

Comment:
Nit: there's a lot of looping and branching in this test definition which impedes readability.
Consider breaking out helper funcs, since the details of these loops don't matter in terms of expressing the intent of the test.

Reply:
I've simplified the test. There was some overly complex logic that boiled down to asserting the same thing twice. Should be better now.